Commands Plugin

randomeizer opened the pull request:
I've created a `hs.commands` module which is used to create to plugins: `commands.global` and `commands.fcpx`. The `global` commands are always on, while the `fcpx` ones get disabled when FCPX is not the active window.

I've ported across all the shortcuts to the other plugins/menus that were already migrated in #217. I'll do the rest in their individual issue branches later.

randomeizer committed:
#246
* Initial checkin of `hs.commands`
* Added `commands.global` and `commands.fcpx` command groups.
* Migrated FCPXHackLaunchFinalCutPro to `fcpx.openfcpx` plugin.

randomeizer committed:
#246
* Migrated configuration for `FCPXHackEffectsX` into the `timeline.effects` plugin
* NOTE: Does not currently support custom shortcuts via the Command Editor.

randomeizer committed:
#246
* Extra documentation
* Tweaks to `hs.commands`

randomeizer committed:
#246
* Tweaks to `fcp:getActiveCommandSet(…)`

randomeizer committed:
Merge branch 'develop' into feature/issue-246

randomeizer committed:
#246
* Wired up switching between default/Command Editor shortcuts into the `hacks.shortcuts` plugin
* BUG: Currently not clearing default shortcuts for commands which are initialised _after_ the `hacks.shortcuts` plugin finishes loading due to how commands are initialised.

randomeizer committed:
#246
* Added support for an optional ‘postInit’ function in plugins which is helpful for remapping shortcuts to FCPX Command Sets.
* Updated `hacks.shortcuts` to use `postInit` to initialise all shortcuts from plugins.

randomeizer committed:
#246
* Migrated ‘Shortcuts’ keyboard shortcuts to ‘hs.commands’ in plugins.

randomeizer committed:
#246
* Renamed ‘whenXXXDo’ functions to ‘whenXXX’ (e.g. `whenPressedDo(…)` is now `whenPressed(…)`

randomeizer committed:
#246
* Migrated `transitionShortcut` shortcuts to commands in the `timeline.transitions` plugin.

randomeizer committed:
#246
* Migrated `effectsShortcut` and `titlesShortcut` to the `timeline.effects` and `timeline.titles` plugins, respectively.

randomeizer committed:
#246
* Migrated `effectsShortcut` and `titlesShortcut` to the `timeline.effects` and `timeline.titles` plugins, respectively.

randomeizer committed:
#246
* Migrated `generatorsShortcut` shortcuts to commands in the `timeline.generators` plugin.

randomeizer committed:
#246
* Migrated `generatorsShortcut` shortcuts to commands in the `timeline.generators` plugin.

randomeizer committed:
#246
* Cleanup of migration of effects/transitions/titles/gererators into plugins.

randomeizer committed:
#246
* Migrated `toggleScrollingTimeline` calls to the `timeline.playhead` plugin.

randomeizer committed:
#246
* Migrated `togglePlayheadLock` to the `timeline.playhead` plugin.

randomeizer committed:
#246
* Migrated `batchExport` and `displayShortcutsList` functions to their respective plugins.

latenitefilms commented:
Error:

```
-----------------------------
| FCPX Hacks v0.79          |
| Created by LateNite Films |
-----------------------------
 > menubar: Loaded menu map from 'hs/finalcutpro/menumap.json' [07:19:21]
Hammerspoon Version:            0.9.52
macOS Version:                  10.12.2
Final Cut Pro Version:          10.3.2
Final Cut Pro Language:         en
FCPX Hacks Locale:              en
Current Keyboard Layout:        U.S.
Final Cut Pro Path:             /Applications/Final Cut Pro.app

 > plugins: Loaded plugin 'hs.fcpxhacks.plugins.menu.manager'
 > Loaded plugin 'hs.fcpxhacks.plugins.menu.bottom'
 > Loaded plugin 'hs.fcpxhacks.plugins.menu.preferences'
 > Loaded plugin 'hs.fcpxhacks.plugins.menu.preferences.menubar'
 > Loaded plugin 'hs.fcpxhacks.plugins.menu.automation'
 > Loaded plugin 'hs.fcpxhacks.plugins.menu.automation.options'
 > wfilter: new hs.window.filter: ... (0x60800086cc80) - reject all with exceptions
 > wfilter: setting default filter: reject
 > wfilter: setting filter for Final Cut Pro: visible=true,
 > added callback for event windowNotOnScreen
 > wfilter: registering 65 running apps
 > wfilter: new app loginwindow registered
 > app universalaccessd has no GUI
 > app SystemUIServer has no GUI
 > app Dock has no GUI
 > wfilter: new app Finder registered
 > found 1 windows for app Finder
 > app Photos Agent has no GUI
 > wfilter: new app ChronoSync Scheduler registered
 > app iTunes Helper has no GUI
 > wfilter: new app Dropbox registered
 > wfilter: new app Google Drive registered
 > wfilter: new app mInstaller registered
 > mInstaller (45) was not registered
 > app com.apple.dock.extra has no GUI
 > wfilter: new app Spotlight registered
 > app Dropbox Finder Integration has no GUI
 > app imklaunchagent has no GUI
 > app AirPlayUIAgent has no GUI [07:19:22]
 > wfilter: new app Notification Center registered
 > wfilter: new app Siri registered
 > app Wi-Fi has no GUI
 > app nbagent has no GUI
 > app FolderActionsDispatcher has no GUI
 > wfilter: new app com.paragon-software.NTFS.fsnotify.agent registered
 > wfilter: new app Desktop Video Firmware Updater registered
 > wfilter: new app Creative Cloud registered
 > wfilter: new app AdobeIPCBroker registered
 > app LaterAgent has no GUI
 > app SafariNotificationAgent has no GUI
 > wfilter: new app Mail registered
 > found 3 windows for app Mail
 > app AppleSpell.service has no GUI
 > app Adobe Desktop Service has no GUI
 > app Mail Web Content has no GUI
 > wfilter: new app Core Sync registered
 > app Touch Bar agent has no GUI
 > app Control Strip has no GUI
 > app Input Menu (DFRExtra) has no GUI
 > app EmojiFunctionRowIM has no GUI
 > app ViewBridgeAuxiliary has no GUI
 > app Spotlight Networking has no GUI
 > app com.apple.appkit.xpc.openAndSavePanelService has no GUI
 > app com.apple.PressAndHold has no GUI
 > app CalNCService has no GUI
 > app storeuid has no GUI
 > wfilter: new app OSDUIHelper registered
 > wfilter: new app NowPlayingTouchUI registered
 > app com.apple.speech.speechsynthesisd has no GUI
 > wfilter: new app GitHub Desktop registered
 > found 1 windows for app GitHub Desktop
 > wfilter: new app universalAccessAuthWarn registered
 > wfilter: new app Terminal registered
 > app CoreServicesUIAgent has no GUI
 > wfilter: new app TextEdit registered
 > wfilter: new app Photoshop registered
 > app dynamiclinkmanager has no GUI
 > wfilter: new app CEPHtmlEngine registered
 > app Dropbox Finder Integration has no GUI
 > app Mail Web Content has no GUI
 > wfilter: new app Final Cut Pro registered
 > found 1 windows for app Final Cut Pro
 > app Codec32BitTool has no GUI
 > wfilter: new app Microsoft Excel registered
 > app Microsoft AU Daemon has no GUI
 > wfilter: new app Hammerspoon registered
 > app Hammerspoon is the frontmost app
 > wfilter: windowfilter instance started (active mode)
 > added Spaces-aware subscriptions
 > added callback for event windowOnScreen
 > wfilter: new hs.window.filter: ... (0x61000066bd00) - custom function
 > added callback for event windowVisible
 > wfilter: windowfilter instance started (active mode)
 > added callback for event windowNotVisible
 > wfilter: new hs.window.filter: ... (0x608000a61fc0) - custom function
 > added callback for event windowVisible
 > wfilter: windowfilter instance started (active mode)
 > added callback for event windowNotVisible
 > plugins: Loaded plugin 'hs.fcpxhacks.plugins.commands.fcpx'
 > wfilter: new hs.window.filter: ... (0x618000863980) - reject all with exceptions
 > wfilter: setting default filter: reject
 > wfilter: setting filter for Final Cut Pro: visible=true,
 > added callback for event windowNotOnScreen
 > wfilter: windowfilter instance started (active mode)
 > added Spaces-aware subscriptions
 > added callback for event windowOnScreen
 > plugins: Loaded plugin 'hs.fcpxhacks.plugins.timeline.playhead'
 > Debug Mode Activated.
 > Loading plugin package 'hs.fcpxhacks.plugins'
 > Loaded plugin 'hs.fcpxhacks.plugins.commands.global'
 > Loaded plugin 'hs.fcpxhacks.plugins.export.batch'
 > Loaded plugin 'hs.fcpxhacks.plugins.menu.top'
 > hotkey: Enabled hotkey ⌘⌃⌥L
 > plugins: Loaded plugin 'hs.fcpxhacks.plugins.fcpx.openfcpx'
 > wfilter: new hs.window.filter: ... (0x618000864a00) - custom function
 > added callback for event windowVisible
 > wfilter: windowfilter instance started (active mode)
 > added callback for event windowNotVisible
 > plugins: Loaded plugin 'hs.fcpxhacks.plugins.fullscreen.shortcuts'
 > Loaded plugin 'hs.fcpxhacks.plugins.hacks.quit'
 > hotkey: Enabled hotkey ⌘⌃⌥F1
 > plugins: Loaded plugin 'hs.fcpxhacks.plugins.hacks.shortcuts'
 > Watching for new media...
 > Loaded plugin 'hs.fcpxhacks.plugins.import.ignorecard'
 > Loaded plugin 'hs.fcpxhacks.plugins.menu.shortcuts'
 > Loaded plugin 'hs.fcpxhacks.plugins.import.preferences'
 > Loaded plugin 'hs.fcpxhacks.plugins.language'
 > Loaded plugin 'hs.fcpxhacks.plugins.timeline.effects'
 > Loaded plugin 'hs.fcpxhacks.plugins.timeline.generators'
 > Loaded plugin 'hs.fcpxhacks.plugins.timeline.preferences'
 > Loaded plugin 'hs.fcpxhacks.plugins.timeline.titles'
 > Loaded plugin 'hs.fcpxhacks.plugins.timeline.transitions'
 > shortcuts: Initialising shortcuts...
 > Applying FCPX Command Editor Shortcuts
 > fcp: Reloading FCPX preferences from file...
 > shortcuts: Applying FCPX Shortcuts to global commands...
 > Looping through all commands...
 > Processing 'FCPXHackShowListOfShortcutKeys'...
 > hotkey: Disabled hotkey ⌘⌃⌥F1
 > hotkey: Deleted hotkey ⌘⌃⌥F1
 > shortcuts: Applying FCPX shortcuts for 'FCPXHackShowListOfShortcutKeys'
 > hotkey: Enabled hotkey ⌘⌃⌥F1
 > shortcuts: Processing 'FCPXHackLaunchFinalCutPro'...
 > hotkey: Disabled hotkey ⌘⌃⌥L
 > hotkey: Deleted hotkey ⌘⌃⌥L
 > shortcuts: Applying FCPX shortcuts for 'FCPXHackLaunchFinalCutPro'
 > hotkey: Enabled hotkey ⌘⌃⌥L
 > shortcuts: Applying FCPX Shortcuts to FCPX commands...
 > Looping through all commands...
 > Processing 'FCPXHackBackgroundRenderOff'...
 > Applying FCPX shortcuts for 'FCPXHackBackgroundRenderOff'
 > *** ERROR: /Users/latenitechris/.hammerspoon/hs/commands/command.lua:111: Invalid key:  - this may mean that the key requested does not exist in your keymap (particularly if you switch keyboard layouts frequently)
stack traceback:
	[C]: in function 'error'
	...oon.app/Contents/Resources/extensions/hs/hotkey/init.lua:23: in upvalue 'getKeycode'
	...oon.app/Contents/Resources/extensions/hs/hotkey/init.lua:190: in function 'hs.hotkey.new'
	/Users/latenitechris/.hammerspoon/hs/commands/command.lua:111: in function 'hs.commands.command._addHotkey'
	/Users/latenitechris/.hammerspoon/hs/commands/command.lua:84: in function 'hs.commands.command.activatedBy'
	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:184: in upvalue 'applyShortcut'
	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:210: in upvalue 'applyCommandShortcut'
	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:219: in upvalue 'applyShortcuts'
	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:233: in upvalue 'applyCommandSetShortcuts'
	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:296: in field 'update'
	...
	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:216: in function 'menuManager'
	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:245: in function 'loadScript'
	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:4327: in main chunk
	[C]: in function 'rawrequire'
	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:449: in function 'require'
	/Users/latenitechris/.hammerspoon/hs/fcpxhacks/init.lua:260: in function </Users/latenitechris/.hammerspoon/hs/fcpxhacks/init.lua:184>
	(...tail calls...)
	[C]: in function 'rawrequire'
	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:449: in function 'require'
	/Users/latenitechris/.hammerspoon/init.lua:4: in main chunk
	[C]: in function 'xpcall'
	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:481: in function 'hs._coresetup.setup'
	(...tail calls...)
 > ** Warning:   wfilter: Hammerspoon (58244) is not registered yet [07:19:22]
 > found 1 windows for app Hammerspoon
 > Hammerspoon (58244) already registered
 > found 1 windows for app Notification Center [07:19:23]
 > ** Warning:   wfilter: No accessibility access to app Mail Networking (no watcher pid) [07:19:27]
 > ** Warning:   wfilter: No accessibility access to app Adobe CEF Helper (no watcher pid) [07:19:27]
 > ** Warning:   wfilter: No accessibility access to app CCXProcess (no watcher pid) [07:19:27]
 > ** Warning:   wfilter: No accessibility access to app CCLibrary (no watcher pid) [07:19:27]
 > ** Warning:   wfilter: No accessibility access to app CEPHtmlEngine Helper (no watcher pid) [07:19:27]
 > wfilter: new app Safari registered [07:19:46]
 > found 1 windows for app Safari
 > found 1 windows for app Safari
 > app Safari is the frontmost app
 > window 58250 is the focused window
 > found 2 windows for app Safari [07:19:52]
 > found 2 windows for app Safari [07:19:53]
```

randomeizer committed:
#246
* Updated `hs.commands` to use the ‘english key code cache’ when creating shortcuts
* Modified `hs.commands` to have shortcuts, which may contain hotkeys, rather than storing hotkeys directly.
* Moved the FCPX-related shortcut functions into the `hs.finalcutpro` project rather than `hs.fcpxhacks.modules.shortcuts.keycodes`.

randomeizer committed:
Merge branch 'develop' into feature/issue-246

# Conflicts:
#	src/hs/plugins/init.lua

randomeizer commented:
Hi @latenitefilms. Can you test again with the latest push?

Currently, not all commands have been migrated to the new system - it's a mix of old-style and new-style creation. An example of one which has been migrated is 'Enable/Disable Scrolling Timeline'. An example of one which has not been migrated is 'Multicam Match Frame'. Basically, if it's still in the `defaultShortcutKeys()` function it's old. If not, it's new.

Could you test that both still work as expected?

latenitefilms committed:
#246

- Updated App Icon
- Added Error Checking to shortcut.lua

latenitefilms commented:
It's still failing for me because of the following error when Hacks Shortcuts are enabled:

```
 > *** ERROR: /Users/latenitechris/.hammerspoon/hs/commands/shortcut.lua:145: Invalid key:  - this may mean that the key requested does not exist in your keymap (particularly if you switch keyboard layouts frequently)
stack traceback:
	[C]: in function 'error'
	...oon.app/Contents/Resources/extensions/hs/hotkey/init.lua:23: in upvalue 'getKeycode'
	...oon.app/Contents/Resources/extensions/hs/hotkey/init.lua:190: in function 'hs.hotkey.new'
	/Users/latenitechris/.hammerspoon/hs/commands/shortcut.lua:145: in function 'hs.commands.shortcut.bind'
	/Users/latenitechris/.hammerspoon/hs/commands/command.lua:130: in function 'hs.commands.command.addShortcut'
	/Users/latenitechris/.hammerspoon/hs/commands/command.lua:115: in function 'hs.commands.command.setShortcuts'
	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:167: in upvalue 'applyShortcuts'
	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:182: in upvalue 'applyCommandSetShortcuts'
	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:245: in field 'update'
	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:267: in field 'init'
	...
	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:245: in function 'loadScript'
	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:4257: in main chunk
	[C]: in function 'rawrequire'
	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:449: in function 'require'
	/Users/latenitechris/.hammerspoon/hs/fcpxhacks/init.lua:297: in function 'loadScriptVersion'
	/Users/latenitechris/.hammerspoon/hs/fcpxhacks/init.lua:275: in function </Users/latenitechris/.hammerspoon/hs/fcpxhacks/init.lua:184>
	(...tail calls...)
	[C]: in function 'rawrequire'
	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:449: in function 'require'
	/Users/latenitechris/.hammerspoon/init.lua:4: in main chunk
	[C]: in function 'xpcall'
	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:481: in function 'hs._coresetup.setup'
	(...tail calls...)
```

I've added in some error checking (`if keyCode ~= nil and keyCode ~= "" then`), which now allows the script to load. However this error now occurs a lot when first starting the script (with Hacks shortcuts activated):

` > ERROR:  shortcut: keyCode was empty. [11:15:37]`

Disabling Hacks Shortcuts, and there's no error messages.

So something must be up with the translating of the Final Cut Pro Command Set values to Hammerspoon Keycodes.

Oh, and I also updated the App Icon.

randomeizer commented:
Hmm. It should be using the 'English' key code set you had set up by
default when actually creating the hotkeys. And I don't get any errors at
all, which is weird...

Do you have any idea which specific shortcuts are failing?
On Thu, 26 Jan 2017 at 10:21 am, Chris Hocking <notifications@github.com>
wrote:

> It's still failing for me because of the following error when Hacks
> Shortcuts are enabled:
>
>  > *** ERROR: /Users/latenitechris/.hammerspoon/hs/commands/shortcut.lua:145: Invalid key:  - this may mean that the key requested does not exist in your keymap (particularly if you switch keyboard layouts frequently)
> stack traceback:
> 	[C]: in function 'error'
> 	...oon.app/Contents/Resources/extensions/hs/hotkey/init.lua:23: in upvalue 'getKeycode'
> 	...oon.app/Contents/Resources/extensions/hs/hotkey/init.lua:190: in function 'hs.hotkey.new'
> 	/Users/latenitechris/.hammerspoon/hs/commands/shortcut.lua:145: in function 'hs.commands.shortcut.bind'
> 	/Users/latenitechris/.hammerspoon/hs/commands/command.lua:130: in function 'hs.commands.command.addShortcut'
> 	/Users/latenitechris/.hammerspoon/hs/commands/command.lua:115: in function 'hs.commands.command.setShortcuts'
> 	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:167: in upvalue 'applyShortcuts'
> 	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:182: in upvalue 'applyCommandSetShortcuts'
> 	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:245: in field 'update'
> 	...is/.hammerspoon/hs/fcpxhacks/plugins/hacks/shortcuts.lua:267: in field 'init'
> 	...
> 	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:245: in function 'loadScript'
> 	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:4257: in main chunk
> 	[C]: in function 'rawrequire'
> 	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:449: in function 'require'
> 	/Users/latenitechris/.hammerspoon/hs/fcpxhacks/init.lua:297: in function 'loadScriptVersion'
> 	/Users/latenitechris/.hammerspoon/hs/fcpxhacks/init.lua:275: in function </Users/latenitechris/.hammerspoon/hs/fcpxhacks/init.lua:184>
> 	(...tail calls...)
> 	[C]: in function 'rawrequire'
> 	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:449: in function 'require'
> 	/Users/latenitechris/.hammerspoon/init.lua:4: in main chunk
> 	[C]: in function 'xpcall'
> 	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:481: in function 'hs._coresetup.setup'
> 	(...tail calls...)
>
> I've added in some error checking (if keyCode ~= nil and keyCode ~= ""
> then), which now allows the script to load. However this error now occurs
> a lot when first starting the script (with Hacks shortcuts activated):
>
> > ERROR: shortcut: keyCode was empty. [11:15:37]
>
> Disabling Hacks Shortcuts, and there's no error messages.
>
> So something must be up with the translating of the Final Cut Pro Command
> Set values to Hammerspoon Keycodes.
>
> Oh, and I also updated the App Icon.
>
> —
> You are receiving this because you authored the thread.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/fcpxhacks/fcpxhacks/pull/252#issuecomment-275274881>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbEw54-iD2S4ENEHhHXy0rvEB0gJhv9ks5rV-cHgaJpZM4LqbKM>
> .
>

latenitefilms commented:
Not sure... couldn't work out how to get that function to "log" which function is was working on. I can have a play later today.

I couldn't see where you've moved `keycodes.translateKeyboardCharacters()`, `keycodes.translateKeyboardKeypadCharacters()`, `keycodes.translateKeyboardModifiers()` and `keycodes.translateModifierMask()` to (it was originally in `fcpxhacks/modules/shortcuts/keycodes.lua`) - so if you're not using these functions anymore, I'd imagine that's the problem?

randomeizer committed:
#246
* Added extra detail about missing key codes in log.

randomeizer commented:
I've moved those functions to the `hs.finalcutpro.keycodes` module. And I had renamed the functions before that for other reasons. They could get renamed back again now that it's in the `hs.finalcutpro` section.

randomeizer commented:
With regards to the actual problem, I've added some extra detail to the log message when the keycode can't be found - it should now also log what character string it was unable to find. That might help in narrowing down the issue.

randomeizer commented:
Could you send me your custom Command Set? There might be some characters you have mapped that I'm not using.

latenitefilms commented:
> I've moved those functions to the hs.finalcutpro.keycodes module.

Ah, ha! No worries! All good - I only had a few minutes to test, so didn't have a good look for them - just noticed they weren't there.

> With regards to the actual problem, I've added some extra detail to the log message when the keycode can't be found - it should now also log what character string it was unable to find. That might help in narrowing down the issue.

Yeah, I already tried checking the `self:getKeyCode()` value - it's just "".

Is there an easy way to work out the "name" of the pressedFn function?

Here's the errors I'm getting:

```
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
 > ** Warning:  shortcut: Unable to find key code for ''. [19:45:14]
```

> Could you send me your custom Command Set? There might be some characters you have mapped that I'm not using.

I'm not using a Custom Command Set - it's just the Default values. However, it also gives me same errors when using a Custom Command Set too.

randomeizer commented:
This is only happening when using the FCPX Command Editor to customise the shortcut, right? It must be some combination of 'characters' that got missed when I moved the code into `hs.finalcutpro`. I'll add a log entry there.

randomeizer committed:
#246
* Added extra logging for command shortcut creation.

randomeizer commented:
Mot logging pushed.

latenitefilms commented:
Maybe it's a non-issue - as none of these should have default shortcuts anyway. We just need some error detection in there ignoring any shortcuts that aren't in the Command Set.

```
> ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTransitionsThree': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTitlesFive': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCreateMulticamOptimizedMediaOn': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTitlesThree': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackLeaveInPlaceOn': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCreateOptimizedMediaOn': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTitlesTwo': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackGeneratorsTwo': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTransitionsTwo': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTitlesOne': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCreateProxyMediaOn': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackBackgroundRenderOff': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackLeaveInPlaceOff': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackGeneratorsFour': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackLockPlayhead': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTransitionsFour': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTransitionsFive': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTransitionsOne': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackGeneratorsFive': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackTitlesFour': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCreateProxyMediaOff': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackBackgroundRenderOn': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackGeneratorsOne': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCreateOptimizedMediaOff': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackGeneratorsThree': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCreateMulticamOptimizedMediaOff': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > Final Cut Pro Window Resized
 > clipboard: Starting Clipboard Watcher.
 > Started Clipboard Watcher
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXCopyWithCustomLabelAndFolder': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle15Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckThreeDown': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneTwo': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneOne': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckFourRight': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle03Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle08Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackToggleVoiceCommands': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneEight': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckOneUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle16Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle14Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle11Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle06Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackPuckOneMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckThreeDown': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckTwoUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle15Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckThreeMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckTwoDown': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle07Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckThree': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckThreeDown': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle01Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneFour': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckTwoRight': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle10Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle10Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle01Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckFourUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckOneRight': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckFourDown': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckFourDown': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackPuckFourMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle03Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackMoveToPlayhead': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckOne': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckOne': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckTwoUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckTwoUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckThree': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckFour': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle16Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle14Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle13Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle12Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle04Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle11Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle09Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckFourMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle13Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle07Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle06Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckThreeUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle05Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckTwoMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle04Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle03Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle02Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle15Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle12Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckOneMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle09Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle10Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckOneUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle08Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckTwoMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle02Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckThreeMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckTwoMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle05Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackPuckThreeMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle16Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckThreeUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckFourDown': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckFour': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle13Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckFourUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle12Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle09Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle07Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckFourMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckFourMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle05Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle04Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckOneUp': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle02Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle08Both': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle01Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckOneMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckOneLeft': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle06Video': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneTen': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckThreeMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneThree': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle11Audio': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackPuckTwoMouse': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneNine': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneFive': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckTwoLeft': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckTwo': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckOneDown': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneSix': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckTwo': {
  characterString = "",
  modifiers = ""
} [20:56:54]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckFourUp': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXAddNoteToSelectedClip': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckTwoDown': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackCutSwitchAngle14Video': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckTwoDown': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckThree': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckOneDown': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSaturationPuckThreeUp': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckThreeRight': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckFourLeft': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckTwo': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckThreeLeft': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackColorPuckOneDown': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckFour': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackChangeSmartCollectionsLabel': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckOne': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackSelectClipAtLaneSeven': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXCopyWithCustomLabel': {
  characterString = "",
  modifiers = ""
} [20:56:55]
 > ** Warning:   fcp: Unable to determine the key code for 'FCPXHackExposurePuckOneMouse': {
  characterString = "",
  modifiers = ""
} [20:56:55]
```

latenitefilms commented:
Also:

```
2017-01-26 21:05:26: ********
 > ERROR:   LuaSkin: hs.application.watcher callback error: ...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:3120: attempt to call a nil value (field 'deleteAllHighlights')
stack traceback:
	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:3120: in function 'deleteAllHighlights'
	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:3468: in function 'finalCutProNotActive'
	/Users/latenitechris/.hammerspoon/hs/finalcutpro/init.lua:1267: in method '_notifyWatchers'
	/Users/latenitechris/.hammerspoon/hs/finalcutpro/init.lua:1235: in function </Users/latenitechris/.hammerspoon/hs/finalcutpro/init.lua:1221> [21:05:26]
2017-01-26 21:05:26: ********
2017-01-26 21:05:28: ********
 > ERROR:   LuaSkin: hs.application.watcher callback error: ...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:3120: attempt to call a nil value (field 'deleteAllHighlights')
stack traceback:
	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:3120: in function 'deleteAllHighlights'
	...nitechris/.hammerspoon/hs/fcpxhacks/modules/fcpx10-3.lua:3468: in function 'finalCutProNotActive'
	/Users/latenitechris/.hammerspoon/hs/finalcutpro/init.lua:1267: in method '_notifyWatchers'
	/Users/latenitechris/.hammerspoon/hs/finalcutpro/init.lua:1235: in function </Users/latenitechris/.hammerspoon/hs/finalcutpro/init.lua:1221> [21:05:28]
2017-01-26 21:05:28: ********
```

randomeizer committed:
#246
* Removed warning log for empty FCPX Commands.

latenitefilms commented:
[This](https://github.com/Hammerspoon/hammerspoon/pull/1194) might also get rid of the need for the English Keycode Map as well if we want to include the yet-to-be-released `hs.keycodes` update.

randomeizer commented:
I'm not sure why you would have gotten the 'deleteAllHighlights' error. I've been moving that around in the #241 issue, but as far as I can tell it's still in #246.
